pydantic_ai_slim/pydantic_ai/models/function.py

```diff
@@ -158,81 +164,83 @@ async def request(
-
-    _next: str | None
-    _iter: AsyncIterator[str]
-    _timestamp: datetime = field(default_factory=_utils.now_utc, init=False)
-    _buffer: list[str] = field(default_factory=list, init=False)
-
-    async def __anext__(self) -> None:
-        if self._next is not None:
-            self._buffer.append(self._next)
-            self._next = None
-        else:
-            self._buffer.append(await self._iter.__anext__())
-
-    def get(self, *, final: bool = False) -> Iterable[str]:
-        yield from self._buffer
-        self._buffer.clear()
-
-    def usage(self) -> result.Usage:
-        return result.Usage()
-
-    def timestamp(self) -> datetime:
-        return self._timestamp
+        yield FunctionStreamedResponse(response_stream)
 
 
 @dataclass
-class FunctionStreamStructuredResponse(StreamStructuredResponse):
-    """Implementation of `StreamStructuredResponse` for [FunctionModel][pydantic_ai.models.function.FunctionModel]."""
+class FunctionStreamedResponse(StreamedResponse):
+    """Implementation of `StreamedResponse` for [FunctionModel][pydantic_ai.models.function.FunctionModel]."""
 
-    _next: DeltaToolCalls | None
-    _iter: AsyncIterator[DeltaToolCalls]
-    _delta_tool_calls: dict[int, DeltaToolCall] = field(default_factory=dict)
+    _iter: AsyncIterator[str | DeltaToolCalls]
     _timestamp: datetime = field(default_factory=_utils.now_utc)
 
-    async def __anext__(self) -> None:
-        if self._next is not None:
-            tool_call = self._next
-            self._next = None
-        else:
-            tool_call = await self._iter.__anext__()
-
-        for key, new in tool_call.items():
-            if current := self._delta_tool_calls.get(key):
-                current.name = _utils.add_optional(current.name, new.name)
-                current.json_args = _utils.add_optional(current.json_args, new.json_args)
+    _next_part_index: int = field(default=0, init=False)
+    _content_part_index: int | None = field(default=None, init=False)
+    _tool_call_index_to_part_index: dict[int, int] = field(default_factory=dict, init=False)
+    _parts: dict[int, ModelResponsePart] = field(default_factory=dict, init=False)
+    _event_iterator: AsyncIterator[ModelResponseStreamEvent] | None = field(default=None, init=False)
+
+    async def __anext__(self) -> ModelResponseStreamEvent:
+        if self._event_iterator is None:
+            self._event_iterator = self._get_event_iterator()
+        return await self._event_iterator.__anext__()
+
+    async def _get_event_iterator(self) -> AsyncIterator[ModelResponseStreamEvent]:
+        async for item in self._iter:
+            # TODO: Create a PartsStreamManager class that wraps
+            #   _next_part_index, _content_part_index, _tool_call_index_to_part_index, and _parts
+            #   and is reusable across the different StreamedResponse implementations.
```

Comment:
I still plan to do this before we merge this code, I think it will help significantly reduce the amount of repetition in the various models in this PR.